### PR TITLE
fix(k8s-events): always attach Node's events to a host

### PR DESCRIPTION
**Pull request by clamoriniere**
### What does this PR do?

Fix how Kubernetes events forwarding detects the Node/Host.
* Previously Nodes' events were not always attached to the correct host.
* Pods' events from "custom" controllers might still be not attached to
  a host if the controller doesn't set the host in the `source.host` event's field.

### Motivation

Only the Kubelet events have the proper `Host` attach to them, because the kubelet
add in `source.host` its Node name.
But when an event is generated by the another controller, this field is empty.

using `involvedObject.name` field for Nodes' events allow to always be able to attach
an Host to the Nodes'events. 

### Additional Notes

Pods' events without `source.host` info are still not assigned to the correct host.
It can be fixed when the workloadmetadatastore is available in the cluster-agent.

Some examples:

* Node event from autoscaler: 
  ```yaml
  apiVersion: v1
  count: 1
  eventTime: null
  firstTimestamp: 2022-11-25T09:40:43Z
  involvedObject:
    apiVersion: v1
    kind: Node
    name: ***************
    resourceVersion: "708461786"
    uid: 4e826fbe-b725-4a06-940a-983d25af0a85
  kind: Event
  lastTimestamp: 2022-11-25T09:40:43Z
  message: failed to drain the node, aborting ScaleDown
  metadata:
    creationTimestamp: 2022-11-25T09:40:43Z
    name: ***************.172aca7b5aa25acb
    namespace: default
    resourceVersion: "708466415"
    uid: de09866b-1cd9-4948-b906-af6e1a58998b
  reason: ScaleDownFailed
  reportingComponent: ""
  reportingInstance: ""
  source:
    component: cluster-autoscaler
  type: Warning
  ```
* kubelet Node event
  ```yaml
  apiVersion: v1
  count: 1
  eventTime: null
  firstTimestamp: 2022-11-25T10:57:27Z
  involvedObject:
    kind: Node
    name: ***************
    uid: ***************
  kind: Event
  lastTimestamp: 2022-11-25T10:57:27Z
  message: invalid capacity 0 on image filesystem
  metadata:
    creationTimestamp: 2022-11-25T10:57:27Z
    name: ***************.172aceab6c0cf73f
    namespace: default
    resourceVersion: "708574550"
    uid: 160131d1-8dd3-446b-8717-734e14086155
  reason: InvalidDiskCapacity
  reportingComponent: ""
  reportingInstance: ""
  source:
    component: kubelet
    host: ***************
  type: Warning
  ```

### Describe how to test/QA your changes

Deploy the cluster-agent in a cluster with event collection enabled.
If you are using the unbundling event option. you will need to allow more event to be
forwarded.

You can create an event attached to a Node with this event template

```yaml
apiVersion: v1
kind: Event
involvedObject:
  kind: Node
  name: "<node-name>"
  uid: "<node-name>"
message: Testing event
metadata:
  name: my-test-event
  namespace: default
reason: Test
source:
  component: my-controller
type: Normal

```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by kayayarai — APPROVED** [review]
👍  from docs

File: pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go [hunk 1]
```diff
@@ -75,14 +75,33 @@ func getInvolvedObjectTags(involvedObject v1.ObjectReference) []string {
 	return tags
 }
 
+const (
+	podKind  = "Pod"
+	nodeKind = "Node"
+)
+
 func getEventHostInfo(clusterName string, ev *v1.Event) eventHostInfo {
+	return getEventHostInfoImpl(getHostProviderID, clusterName, ev)
```
Comment (vboulineau) [inline pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go:84]: Why do we need two functions? If it's only for unit testing, we should have a comment.

Comment (clamoriniere) [reply pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go:84]: it is for unit-test. I didn't wanted to call the code that retrieve the host info

File: pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go [hunk 2]
```diff
@@ -75,14 +75,33 @@ func getInvolvedObjectTags(involvedObject v1.ObjectReference) []string {
...
+}
+
+func getEventHostInfoImpl(hostProviderIDFunc func(string) string, clusterName string, ev *v1.Event) eventHostInfo {
```
Comment (clamoriniere) [inline pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go:87]:
```suggestion
// getEventHostInfoImpl get the event host information (hostname,nodename)
// this function takes `hostProviderIDFunc` function to ease unit-testing by mocking the
// providers logic
func getEventHostInfoImpl(hostProviderIDFunc func(string) string, clusterName string, ev *v1.Event) eventHostInfo {
```

**Review by vboulineau — APPROVED** [review]

**Pull request merged**